octave/ofdm_lib.m

```diff
@@ -388,11 +433,34 @@ function print_config(states)
   states.rxbuf(1:Nrxbuf-states.nin) = states.rxbuf(states.nin+1:Nrxbuf);
   states.rxbuf(Nrxbuf-states.nin+1:Nrxbuf) = rxbuf_in;
 
-  % Attempt coarse timing estimate (i.e. detect start of frame)
+  % Attempt coarse timing estimate (i.e. detect start of frame) at a range of frequency offsets
 
-  st = M+Ncp + Nsamperframe + 1; en = st + 2*Nsamperframe; 
-  [ct_est timing_valid timing_mx] = est_timing(states, states.rxbuf(st:en), states.rate_fs_pilot_samples);
-  [foff_est states] = est_freq_offset(states, states.rxbuf(st:en), states.rate_fs_pilot_samples, ct_est);
+  st = M+Ncp + Nsamperframe + 1; en = st + 2*Nsamperframe;
+  timing_mx = 0; fcoarse = 0; timing_valid = 0; 
+  for afcoarse=-40:40:40
+    % vector of local oscillator samples to shift input vector
+    % these could be computed on the fly to save memory, or pre-computed in flash at tables as they are static
+    w = 2*pi*afcoarse/Fs;
+    wvec = exp(-j*w*(0:2*Nsamperframe));
+
+    % choose best timing offset metric at this freq offset
+    [act_est atiming_valid atiming_mx] = est_timing(states, wvec .* states.rxbuf(st:en), states.rate_fs_pilot_samples);
+    %printf("afcoarse: %f atiming_mx: %f\n", afcoarse, atiming_mx);
+    
+    if atiming_mx > timing_mx
+      ct_est = act_est;
+      timing_valid = atiming_valid;
+      timing_mx = atiming_mx;
+      fcoarse = afcoarse;
+    end
+  end
+  
```

Comment:
@srsampson @donmr I hope this runs OK in real time on the stm32.  Many opportunities for optimisation, e.g. recursions or precomputation at compile time of ```wvec``` above.